### PR TITLE
Better error message when trying to re-initialize CUDA in forked subprocess

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #14631
Follow up to #13405

We recently launched DDP Fork support in notebooks (ddp_notebook), and some users will soon report issues when seeing this error message in their notebooks: "Cannot re-initialize CUDA in forked subprocess". 

The issue is mitigated by #14631 for our internal calls, and pytorch addresses torch.cuda utility function calls in https://github.com/pytorch/pytorch/pull/84879 and https://github.com/pytorch/pytorch/pull/85024. However, it can always happen that the user allocates CUDA memory accidentally without knowing that this is not supported with the fork start method. We point it out in our documentation, but that's not enough. 

This PR introduces an error message that is more helpful than the default PyTorch one. 

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @akihironitta

### CODE REVIEW

File: src/lightning_lite/strategies/launchers/multiprocessing.py [hunk 1]
```diff
@@ -176,3 +179,22 @@ def restore(self) -> None:
...
+
+    The error message replaces PyTorch's 'Cannot re-initialize CUDA in forked subprocess' with helpful advice for
+    Lightning users.
+    """
+    if not torch.cuda.is_initialized():
+        return
+
+    message = (
+        "Lightning can't create new processes if CUDA is already initialized. Did you manually call"
+        " `torch.cuda.*` functions, have moved the model to the device or allocated memory on the GPU any"
```
Comment (carmocca) [inline src/lightning_lite/strategies/launchers/multiprocessing.py:195]: oxford comma :tophat: 

```suggestion
        " `torch.cuda.*` functions, have moved the model to the device, or allocated memory on the GPU any"
```

File: src/lightning_lite/strategies/launchers/multiprocessing.py [hunk 2]
```diff
@@ -176,3 +179,22 @@ def restore(self) -> None:
...
+        " other way? Please remove any such calls, or change the selected strategy."
+    )
+    if _IS_INTERACTIVE:
+        message += " You will have to restart the Python session; in a notebook, that means restart the kernel."
```
Comment (carmocca) [inline src/lightning_lite/strategies/launchers/multiprocessing.py:199]: What is interactive but not a notebook?

```suggestion
        message += " You will have to restart the Python kernel."
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 3]
```diff
@@ -40,6 +40,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `WandbLogger.download_artifact` and `WandbLogger.use_artifact` for managing artifacts with Weights and Biases ([#14551](https://github.com/Lightning-AI/lightning/issues/14551))
 
 
+- Added a friendlier error message when attempting to fork processes with pre-initialized CUDA context ([#14709](https://github.com/Lightning-AI/lightning/issues/14709))
```
Comment (carmocca) [inline src/pytorch_lightning/CHANGELOG.md:43]:
```suggestion
- Added a more descriptive error message when attempting to fork processes with pre-initialized CUDA context ([#14709](https://github.com/Lightning-AI/lightning/issues/14709))
```

**Review by carmocca — APPROVED** [review]
I am subscribed to your PRs, reviewing them before you even mark them as ready :racehorse:

File: src/lightning_lite/strategies/launchers/multiprocessing.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/launchers/multiprocessing.py:199]: Yes exactly. Maybe I should leave it out and just shorten the message to mentioning notebook?

I distinctly remember someone asking on Slack questions about trainer in the interactive shell. But yes, that's not a serious way to interact with Lightning usually.

**Comment by awaelchli** [comment]
> I am subscribed to your PRs, reviewing them before you even mark them as ready 🐎

Thank you! Is that a horse?

Will mark it ready once I gave this a final check in Google Colab as well.

**Review by akihironitta — APPROVED** [review]
LGTM!

**Review by otaj — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**